pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -29,188 +33,339 @@ type DockerRegistryServiceControllerOptions struct {
-	e.dockercfgController = options.DockercfgController
-	e.defaultDockerURL = options.DefaultDockerURL
+	e.servicesSynced = e.serviceController.HasSynced
+	e.syncRegistryLocationHandler = e.syncRegistryLocationChange
+
+	dockercfgOptions := kapi.ListOptions{FieldSelector: fields.SelectorFromSet(map[string]string{kapi.SecretTypeField: string(kapi.SecretTypeDockercfg)})}
+	e.secretCache, e.secretController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(opts kapi.ListOptions) (runtime.Object, error) {
+				return e.client.Secrets(kapi.NamespaceAll).List(dockercfgOptions)
+			},
+			WatchFunc: func(opts kapi.ListOptions) (watch.Interface, error) {
+				return e.client.Secrets(kapi.NamespaceAll).Watch(dockercfgOptions)
+			},
+		},
+		&kapi.Secret{},
+		options.Resync,
+		framework.ResourceEventHandlerFuncs{},
+	)
+	e.secretsSynced = e.secretController.HasSynced
+	e.syncSecretHandler = e.syncSecretUpdate
 
 	return e
 }
 
 // DockerRegistryServiceController manages ServiceToken secrets for Service objects
 type DockerRegistryServiceController struct {
-	stopChan chan struct{}
-
 	client client.Interface
 
-	registryServiceName string
-	defaultDockerURL    string
+	serviceName      string
+	serviceNamespace string
 
 	dockercfgController *DockercfgController
 
-	serviceController *framework.Controller
+	serviceController           *framework.Controller
+	serviceCache                cache.Store
+	servicesSynced              func() bool
+	syncRegistryLocationHandler func(key string) error
+
+	secretController  *framework.Controller
+	secretCache       cache.Store
+	secretsSynced     func() bool
+	syncSecretHandler func(key string) error
+
+	registryURLs          sets.String
+	registryURLLock       sync.RWMutex
+	registryLocationQueue workqueue.RateLimitingInterface
+	secretsToUpdate       workqueue.RateLimitingInterface
+
+	dockerURLsIntialized chan struct{}
 }
 
 // Runs controller loops and returns immediately
-func (e *DockerRegistryServiceController) Run() {
-	if e.stopChan == nil {
-		e.stopChan = make(chan struct{})
-		go e.serviceController.Run(e.stopChan)
+func (e *DockerRegistryServiceController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	go e.serviceController.Run(stopCh)
+	go e.secretController.Run(stopCh)
+	go e.waitForDockerURLs(stopCh)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(e.watchForDockercfgSecretUpdates, time.Second, stopCh)
 	}
+
+	<-stopCh
+	glog.Infof("Shutting down docker registry service controller")
+	e.registryLocationQueue.ShutDown()
 }
 
-// Stop gracefully shuts down this controller
-func (e *DockerRegistryServiceController) Stop() {
-	if e.stopChan != nil {
-		close(e.stopChan)
-		e.stopChan = nil
-	}
+// enqueue adds to our queue.  We only have one entry, but we never have to check it since we already know the things
+// we're watching for.
+func (e *DockerRegistryServiceController) enqueueRegistryLocationQueue() {
+	e.registryLocationQueue.Add("check")
```

Comment:
Ok, now I see. You don't care about the key but about what it triggers (syncRegistryLocationChange)